### PR TITLE
ARROW-14672: [Docs] Document how to exchange data between Python and Java

### PR DESCRIPTION
**Pull request by amol-**

### CODE REVIEW

File: docs/source/python/integration/python_java.rst [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+:ref:`c-data-interface`.
+
+This can be used to exchange data between Python and Java functions and
+methods so that the two languages can interact without any cost of
+marshaling and unmarshaling data.
+
+.. note::
+
+    The article takes for granted that you have a ``Python`` environment
+    with ``pyarrow`` correctly installed and an ``Java`` environment with
```
Comment (raulcd) [inline docs/source/python/integration/python_java.rst:31]: Minor typo
```suggestion
    with ``pyarrow`` correctly installed and a ``Java`` environment with
```

File: docs/source/python/integration/python_java.rst [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+.. code-block:: bash
+
+    $ pip install jpype1
+
+The most basic thing we can do with our ``Simple`` class is to
+use the ``Simple.getNumber`` method from Python and see 
+if it will return the result.
+
+To do so we can create an ``simple.py`` file which uses ``jpype`` to
```
Comment (raulcd) [inline docs/source/python/integration/python_java.rst:70]:
```suggestion
To do so we can create a ``simple.py`` file which uses ``jpype`` to
```

File: docs/source/python/integration/python_java.rst [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    array = pa.array([0]*10)
+
+    from pyarrow.cffi import ffi as arrow_c
+
+    # Export the Python array through C-Data
+    c_array = arrow_c.new("struct ArrowArray*")
+    c_array_ptr = int(arrow_c.cast("uintptr_t", c_array))
+    array._export_to_c(c_array_ptr)
+
+    # Export the Schema of the Arrayo through C-Data
```
Comment (raulcd) [inline docs/source/python/integration/python_java.rst:340]:
```suggestion
    # Export the Schema of the Array through C-Data
```

**Review by raulcd — COMMENTED** [review]
I took a look and being the first time I have been exposed to this functionality I found it quite clear. I just found a couple minor typos / grammar.

File: docs/source/python/integration/python_java.rst [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+    import jpype.imports
+    from jpype.types import *
+
+    # Init the JVM and make FillTen class available to Python.
+    jpype.startJVM(classpath=["./dependencies/*", "./target/*"])
+    FillTen = JClass('FillTen')
+
+    # Create a Python array of 10 elements
+    import pyarrow as p
+    array = pa.array([0]*10)
```
Comment (davisusanibar) [inline docs/source/python/integration/python_java.rst:331]: pa.array -> p.array

File: docs/source/python/integration/python_java.rst [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+
+Java to Python communication using the C Data Interface
+-------------------------------------------------------
+
+The C-Data interface is a protocol implemented in Arrow to exchange data within different
+environments without the cost of marshaling and copying data.
+
+This allows to expose data coming from Python or Java to functions that are implemented
+in the other language.
+
```
Comment (davisusanibar) [inline docs/source/python/integration/python_java.rst:305]: Add or validate if cffi module is already installed: `pip install cffi`

File: docs/source/python/integration/python_java.rst [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+Now that we have the package built, we can make it available to Python.
+To do so, we need to make sure that not only the package itself is available,
+but that also its dependencies are too.
+
+We can use ``maven`` to collect all dependencies and make them available in a single place
+(the ``dependencies`` directory) so that we can more easily load them from Python:
+
+.. code-block:: bash
+
+    $ mvn org.apache.maven.plugins:maven-dependency-plugin:2.7:copy-dependencies -DoutputDirectory=dependencies
```
Comment (pitrou) [inline docs/source/python/integration/python_java.rst:218]: I'm clueless about Java, but can't we just produce a single jar and give it to jpype?

File: docs/source/python/integration/python_java.rst [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+
+The C-Data interface is a protocol implemented in Arrow to exchange data within different
+environments without the cost of marshaling and copying data.
+
+This allows to expose data coming from Python or Java to functions that are implemented
+in the other language.
+
+.. note::
+
+    In the future the ``pyarrow.jvm`` will be implemented to leverage the C-Data
```
Comment (pitrou) [inline docs/source/python/integration/python_java.rst:308]: "C Data Interface" (no hyphen)

File: docs/source/python/integration/python_java.rst [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+    interface, at the moment is instead specifically written for JPype
+
+To showcase how C-Data works, we are going to tweak a bit both our ``FillTen`` Java
```
Comment (pitrou) [inline docs/source/python/integration/python_java.rst:311]: Still no hyphen (also below :-)).

File: docs/source/python/integration/python_java.rst [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+
+.. code-block:: bash
+
+    $ pip install jpype1
+
+The most basic thing we can do with our ``Simple`` class is to
+use the ``Simple.getNumber`` method from Python and see 
+if it will return the result.
+
+To do so we can create a ``simple.py`` file which uses ``jpype`` to
```
[the remark this answers is not in the document]
Comment (amol-) [reply docs/source/python/integration/python_java.rst:70]: 👍

File: docs/source/python/integration/python_java.rst [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+This can be used to exchange data between Python and Java functions and
+methods so that the two languages can interact without any cost of
+marshaling and unmarshaling data.
+
+.. note::
+
+    The article takes for granted that you have a ``Python`` environment
+    with ``pyarrow`` correctly installed and a ``Java`` environment with
+    ``arrow`` library correctly installed. 
+    The ``Java`` version must have been compiled with ``mvn -Parrow-c-data`` to
```
[the remark this answers is not in the document]
Comment (amol-) [reply docs/source/python/integration/python_java.rst:33]: I think it's ok, it's a more general statement. Independently from _where_ you got your arrow java, it must have been compiled with c-data enabled

File: docs/source/python/integration/python_java.rst [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+    ]
+
+The first line is the raw result of invoking the ``FillTen.createArray`` method.
+The resulting object is a proxy to the actual Java object, so it's not really a pyarrow
+Array, it will lack most of its capabilities and methods. 
+That's why we subsequently use ``pyarrow.jvm.array`` to convert it to an actual
+``pyarrow`` array. That allows us to treat it like any other ``pyarrow`` array.
+The result is the second line in the output where the array is correctly reported
+as being of type ``pyarrow.lib.Int64Array`` and is printed using the ``pyarrow`` style.
+
```
[the remark this answers is not in the document]
Comment (amol-) [reply docs/source/python/integration/python_java.rst:296]: Added a note at the end of the ``pyarrow.jvm`` section

File: docs/source/python/integration/python_java.rst (hunk 4 above)
Comment (amol-) [reply docs/source/python/integration/python_java.rst:331]: :+1:

File: docs/source/python/integration/python_java.rst (hunk 5 above)
Comment (amol-) [reply docs/source/python/integration/python_java.rst:311]: :+1:

File: docs/source/python/integration/python_java.rst [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+            </dependency>
+        </dependencies>
+    </project>
+
+Once the ``FillTen.java`` file with the class is created
+as ``src/main/java/FillTen.java`` we can use ``maven`` to
+compile the project with ``mvn package`` and get it 
+available in the ``target`` directory.
+
+.. code-block:: bash
```
[the remark this answers is not in the document]
Comment (amol-) [reply docs/source/python/integration/python_java.rst]: good catch, changed it.

File: docs/source/python/integration/python_java.rst [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+
+        public static ValueVector createArray() {
+            BigIntVector intVector = new BigIntVector("ints", allocator);
+            intVector.allocateNew(10);
+            intVector.setValueCount(10);
+            FillTen.fillValueVector(intVector);
+            return intVector;
+        }
+
+        private static void fillValueVector(ValueVector v) {
```
[the remark this answers is not in the document]
Comment (amol-) [reply docs/source/python/integration/python_java.rst:121]: Fair point, I originally did it to suggest that you can pass around any kind of ``ValueVector``, but in the end the example depends on it being a fixed width one and specifically uses ``setSafe`` which is implemented by the subclass. I'll change it to ``BigIntVector``

File: docs/source/python/integration/python_java.rst (hunk 6 above)
Comment (amol-) [reply docs/source/python/integration/python_java.rst:216]: Yes, we can using a maven plugin. But I wanted to make explicit the step where you need to collect and load dependencies by yourself. I'll add a note that the alternative is to use `maven-assembly-plugin` to assemble a single jar with all dependencies inside.

**Pull request closed without merging**